src/Http/Http.Api.Abstractions/src/IResult.cs

```diff
@@ -0,0 +1,20 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Http.Api.Abstractions
+{
+    /// <summary>
+    /// Defines a contract that represents the result of an HTTP endpoint.
+    /// </summary>
+    public interface IResult
+    {
+        /// <summary>
+        /// Write an HTTP response reflecting the result.
+        /// </summary>
+        /// <param name="httpContext">The <see cref="HttpContext"/> for the current request.</param>
+        /// <returns>A task that represents the asynchronous execute operation.</returns>
+        ValueTask WriteResponseAsync(HttpContext httpContext);
```

Comment:
We should have a discussion about Task vs ValueTask, I'm not sure this should be ValueTask.

Reply:
This should be `ExecuteAsync` not `WriteResponseAsync`. It might not write any response at all.

Reply:
I'm on the fence about ValueTask. On one hand, RequestDelegate doesn't return a ValueTask and we're not returning an async value so we can just used cached completed tasks were appropriate. On the other hand, ValueTask gives us the most flexibility if we ever want to try anything crazy like returning pooled tasks for async operations.

What potential IResults are you thinking of that don't write a response? I know StatusCodeResult for example doesn't flush the response so something could completely change the response afterwards, but in practice it's effectively writing a response with the specified status code and an empty body the vast majority of the time.

I know technically an IResult could do nothing or anything, but I think naming the method after what it does the majority of the time makes sense. Are there use cases for IResult that have nothing to do with the response that I'm not considering?

Reply:
> I'm on the fence about ValueTask. On one hand, RequestDelegate doesn't return a ValueTask and we're not returning an async value so we can just use Task.CompletedTask were appropriate. On the other hand, ValueTask gives us the most flexibility if we ever want to try anything crazy like returning pooled tasks for async operations. Of course, that's only worthwhile if whoever is awaiting the action doesn't immediately convert the ValueTask to a Task like the middleware pipeline would.

I think it should be `Task` since all of our existing APIs today are Task based. No point introducing friction here.

> What potential IResults are you thinking of that don't write a response? I know StatusCodeResult for example doesn't flush the response so something could completely change the response afterwards, but in practice it's effectively writing a response with the specified status code and an empty body the vast majority of the time.

I think keeping this close to what action results look like is a good thing. It also takes the HttpContext not the HttpResponse. On the pedantic side, writing to the response sounds like it's writing to the body.